File: nodekit/_internal/ops/calculate_bonus.py
import warnings
from decimal import Decimal
from typing import List

from nodekit._internal.types.events.events import Event, EventTypeEnum, NodeResultEvent
from nodekit._internal.types.node import Timeline


# %%
def calculate_bonus_usd(
        events: List[Event],
        timeline: Timeline,
) -> Decimal:
    """
    Calculates the current bonus USD earned based on the provided events.
    """

    calculated_amount = Decimal('0')

    # Sort events by timestamp
    events = sorted(events, key=lambda ev: ev.timestamp_event)
    observed_node_indices = set()
    for ev in events:
        if ev.event_type != EventTypeEnum.NodeResultEvent:
            continue

        ev: NodeResultEvent

        node_index = ev.node_index
        sensor_index = ev.sensor_index

        # Skip if there was already an Event observed for this Node
        if node_index in observed_node_indices:
            warnings.warn(f"Multiple NodeResultEvents observed for Timeline.nodes[{node_index}]. Only the first will be considered for bonus calculation.")
            continue

        sensor = timeline.nodes[node_index].sensors[sensor_index]

        outcome = sensor.outcome
        if outcome is None:
            continue

        calculated_amount += Decimal(outcome.bonus_amount_usd)
        observed_node_indices.add(node_index)

    # Clip at minimum of 0:
    if calculated_amount < Decimal('0'):
        calculated_amount = Decimal('0.00')

    return calculated_amount

File: nodekit-browser/src/ops/calculate-bonus.ts
import type {Event, NodeIndex, SensorIndex} from "../types/events";
import type {Timeline} from "../types/node.ts";

export function calculateBonusUsd(
    events: Event[],
    timeline: Timeline,
): number {
    /*
    Function which computers the bonus USD amount based on the events and the node graph.
     */

    let bonusComputed = 0;

    // Sort events by timestamp to ensure correct order of processing
    // Note that event_timestamp is an ISO8601 string:
    events.sort((a, b) => a.timestamp_event.localeCompare(b.timestamp_event));

    let observedNodes = new Set<NodeIndex>();
    for (let i = 0; i < events.length; i++) {

        const eventCur = events[i];
        // Type narrow to NodeResultEvent:
        if (eventCur.event_type !== 'NodeResultEvent') {
            continue;
        }
        const nodeIndex: NodeIndex = eventCur.node_index;
        const sensorIndex: SensorIndex = eventCur.sensor_index;
        // Skip if we've already processed this node
        if (observedNodes.has(nodeIndex)) {
            continue;
        }
        observedNodes.add(nodeIndex);

        // Get Outcome for this Sensor fire, if it exists:
        const node = timeline.nodes[nodeIndex];
        const sensor = node.sensors[sensorIndex];
        const outcome = sensor.outcome;
        if (!outcome){
            continue
        }
        let bonusAmountUsd = parseFloat(outcome.bonus_amount_usd);
        if (!isNaN(bonusAmountUsd) && bonusAmountUsd > 0){
            bonusComputed += bonusAmountUsd;
        }
    }

    bonusComputed = Math.max(0, bonusComputed);
    bonusComputed = Math.round(bonusComputed * 100) / 100; // Round to 2 decimals

    // Convert to string:
    return bonusComputed;
}